In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import accuracy_score  
from sklearn.metrics import recall_score  
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.linear_model import LogisticRegression 
from sklearn import neighbors
from sklearn import tree
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import BernoulliNB
import matplotlib.pyplot as plt

## get the X_date, Y_target

In [2]:
def get_the_predicted_bugs(name, num1, num2):
    All_Data = pd.read_csv(name, usecols = [i for i in range(num1, num2)]).values
    return All_Data

In [3]:
data = get_the_predicted_bugs('modified_all_data.csv', 2, 45)

## If we predict all the versions are without bugs

In [4]:
accuracy_score_list = []
recall_score_list = []
for i in range(100):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    accuracy_score_list.append(accuracy_score(y_test, [0] * len(y_test)))
    recall_score_list.append(recall_score(y_test, [0] * len(y_test)))

In [5]:
np.mean(accuracy_score_list)

0.8583806343906509

In [6]:
np.mean(recall_score_list)

0.0

If we predict all the versions are without bugs, the average accuracy_score of 100 times is about is about 0.86 and the recall_score is always equal to 0.

## Naive Bayes

## BernoulliNB

In [7]:
accuracy_score_list = []
recall_score_list = []
for i in range(100):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    mnb_clf = BernoulliNB()
    mnb_clf.fit(x_train, y_train)
    
    y_pre = mnb_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [8]:
np.mean(accuracy_score_list)

0.8261268781302169

In [9]:
np.mean(recall_score_list)

0.34259646477055905

By BernoulliNB, the average accuracy_score of 100 times is about is about 0.826 and the average recall_score of 100 times is about 0.34.

## GaussianNB

In [10]:
accuracy_score_list = []
recall_score_list = []
for i in range(100):
    data_random = np.random.permutation(data)
    x = data_random[:, [i for i in range(data.shape[1] - 1)]]
    y = data_random[:, -1]
    x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.4, random_state = 0)
    gnb_clf = GaussianNB()
    gnb_clf.fit(x_train, y_train)
    
    y_pre = gnb_clf.predict(x_test)
    accuracy_score_list.append(accuracy_score(y_test, y_pre))
    recall_score_list.append(recall_score(y_test, y_pre))

In [11]:
np.mean(accuracy_score_list)

0.8318864774624373

In [12]:
np.mean(recall_score_list)

0.3966138236042437

By GaussianNB, the average accuracy_score of 100 times is about is about 0.83 and the average recall_score of 100 times is about 0.40.